keras_hub/src/models/sam/sam_layers.py

```diff
@@ -0,0 +1,396 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+
+import keras
+from keras import ops
+
+
+class MLP(keras.layers.Layer):
+    """A MLP block with architecture.
+
+    `input_dim -> [hidden_dim] * (num_layers - 1) -> output_dim`.
+
+    Args:
+        hidden_dim (int): The number of units in the hidden layers.
+        output_dim (int): The number of units in the output layer.
+        num_layers (int): The total number of dense layers to use.
+        activation (str): Activation to use in the hidden layers.
+            Default is `"relu"`.
+    """
+
+    def __init__(
+        self, hidden_dim, output_dim, num_layers, activation="relu", **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.hidden_dim = hidden_dim
+        self.output_dim = output_dim
+        self.num_layers = num_layers
+        self.activation = activation
+        h = [hidden_dim] * (num_layers - 1)
+        self.mlp_block = []
+        for hidden_dim in h:
+            self.mlp_block.append(
+                keras.layers.Dense(hidden_dim, dtype=self.dtype_policy)
+            )
+            self.mlp_block.append(
+                keras.layers.Activation(activation, dtype=self.dtype_policy)
+            )
+        self.mlp_block.append(
+            keras.layers.Dense(output_dim, dtype=self.dtype_policy)
+        )
+        self.mlp_block = keras.models.Sequential(self.mlp_block)
+
+    def build(self, input_shape):
+        self.mlp_block.build(input_shape)
+        self.built = True
+
+    def call(self, x):
+        return self.mlp_block(x)
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "hidden_dim": self.hidden_dim,
+                "output_dim": self.output_dim,
+                "num_layers": self.num_layers,
+                "activation": self.activation,
+            }
+        )
+        return config
+
+
+class MultiHeadAttentionWithDownsampling(keras.layers.Layer):
+    """Multi-Head Attention with downsampling.
+
+    An attention layer that allows for downscaling the size of the embedding
+    after projection to queries, keys, and values.
+    This layer first downscales the features of input queries, keys, and
+    values using a dense layer. Multi-head attention is then performed
+    and the attention map is projected back (upscaled) to the number of
+    input features.
+
+    Args:
+        num_heads (int): Number of attention heads.
```

Comment:
It'd be great to have a uniform format for how we specify the type of the arguments. For example [here](https://github.com/keras-team/keras-hub/pull/1847/files#diff-2a20cb5cae4196c0d6127ba0b676d3e12a80a199903b130b759f27f70559c029R183) is different.
I think we usually use `argument: type.` format.